### DISCUSSION TITLE
Get remote URL in beforeError hook

### DISCUSSION THREAD

**Discussion by haapti** [sindresorhus/got#1919] [Q&A] [answered]
Is there a way to get remote URL in [beforeError](https://github.com/sindresorhus/got/blob/main/documentation/9-hooks.md#beforeerror) hook without adding [context](https://github.com/sindresorhus/got/discussions/1544) for each request?

For example, on timeout would be handy to log remote URL what caused the timeout.

**Comment by sindresorhus** [comment, marked as answer]
the URL is available on the error object directly. no need to add context manually.

```js
import got from 'got';

const client = got.extend({
	hooks: {
		beforeError: [
			error => {
				// the URL is always available here
				const url = error.options.url?.toString();
				console.log('Request failed for:', url);
				
				// if theres a response, you can also get:
				// - original URL: error.response?.requestUrl
				// - final URL (after redirects): error.response?.url
				
				return error;
			}
		]
	}
});
```

this hook applies to all requests made with the extended client, so you dont need to set it for each request individually.
